File: Fish.cpp
// you can use includes, for example:
// #include <algorithm>
#include <deque>
#include <vector>
#include <utility>
// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(vector<int> &A, vector<int> &B) {
    // write your code in C++11 (g++ 4.8.2)
    deque< pair<int,int> > river;
    int lens = A.size();
    for (int i = 0; i < lens; i++)
    {
        if (river.size() == 0)
        {
            river.push_back( make_pair(A[i],B[i]) );
        }else{
            while (!river.empty())
            {
                auto fish = river.back();
                if (fish.second == 1 && B[i]==0)
                {
                    if (fish.first < A[i])
                    {
                        river.pop_back(); //fish existing is eaten
                        continue;
                    }else{
                        break; //eat the comming fish
                    }
                }else{
                    river.push_back(make_pair(A[i],B[i]));
                    break;//both keep alive
                }
            }
            if (river.empty())
                river.push_back( make_pair(A[i],B[i]) );            
        }
    }
    return river.size();
}

File: Stacks and Queues/Fish.py
# you can write to stdout for debugging purposes, e.g.
# print "this is a debug message"

def solution(A, B):
    # write your code in Python 2.7
    lenth = len(A)
    river = []
    for i in range(lenth):
        if len(river) == 0:
            river.append((A[i],B[i]))
        else:
            while len(river)>0:
                if river[-1][1] == 1 and B[i] == 0:
                    if river[-1][0] < A[i]:
                        river.pop()
                        continue
                    else:
                        break
                else:
                    river.append((A[i],B[i]))
                    break
            else:
                river.append((A[i],B[i]))
    
    return len(river)